src/mame/drivers/uknc.cpp

```diff
@@ -9,97 +9,946 @@ Graphics 640x288 pixels.
 	uint32_t screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect);
-	required_device<k1801vm2_device> m_maincpu;
+
+	DECLARE_WRITE_LINE_MEMBER(screen_vblank);
+	TIMER_DEVICE_CALLBACK_MEMBER(scanline_callback);
 
 	void uknc_mem(address_map &map);
 	void uknc_sub_mem(address_map &map);
+	void uknc_mem_banked(address_map &map);
+	void uknc_cart_mem(address_map &map);
+	void uknc_rom_mem(address_map &map);
+
+	uint16_t trap_r();
+	void trap_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+
+	uint16_t maincpu_vram_addr_r();
+	void maincpu_vram_addr_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t maincpu_vram_data_r();
+	void maincpu_vram_data_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+
+	uint16_t bcsr_r();
+	void bcsr_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t scsr_r();
+	void scsr_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+
+	uint16_t vram_addr_r();
+	void vram_addr_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t vram_data_pl0_r();
+	void vram_data_pl0_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t vram_data_p12_r();
+	void vram_data_p12_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t sprite_r(offs_t offset);
+	void sprite_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+
+	// subcpu registers
+	uint16_t m_bcsr; // 177054
+	uint16_t m_scsr; // 177716
+
+	// video
+	struct
+	{
+		uint16_t vram_addr;
+		uint16_t vram_data_pl0;
+		uint16_t vram_data_p12;
+		uint16_t maincpu_vram_addr;
+		uint16_t maincpu_vram_data;
+		uint8_t foreground;  // 177016
+		uint32_t background; // 177020
+		uint8_t sprite_mask; // 177024
+		uint8_t plane_mask;  // 177026
+		uint32_t control;
+		uint32_t palette;
+	} m_video;
+	std::unique_ptr<uint8_t[]> m_videoram;
+	std::unique_ptr<scanline_t[]> m_scanlines;
+	bitmap_rgb32 m_tmpbmp;
+
+	void uknc_palette(palette_device &palette) const;
+	void set_palette_yrgb(uint32_t aux);
+	void set_palette_ygrb(uint32_t aux);
+	void set_palette__rgb(uint32_t aux);
+	void set_palette_gray(uint32_t aux);
+	void update_displaylist();
+	void draw_scanline(uint32_t *p, scanline_t *scanline);
+
+protected:
+	required_device<k1801vm2_device> m_maincpu;
+	required_device<k1801vm2_device> m_subcpu;
+	required_device<address_map_bank_device> m_bankdev0;
+	required_device<address_map_bank_device> m_bankdev1;
+	required_device<address_map_bank_device> m_bankdev2;
+	required_device<qbus_device> m_qbus;
+	required_device<qbus_device> m_subqbus;
+	required_device<cassette_image_device> m_cassette;
+	required_ioport m_yrgb;
+	required_device<palette_device> m_palette;
+	required_device<screen_device> m_screen;
 };
 
 
 void uknc_state::uknc_mem(address_map &map)
 {
-	map.unmap_value_high();
-	map(0x0000, 0x7fff).ram();
-	map(0x8000, 0xffff).rom().region("maincpu",0);
+	map(0000000, 0177777).m(m_bankdev0, FUNC(address_map_bank_device::amap16));
+}
```

Comment:
You should really use views for this – it’s a lot move efficient than going through the bankdev trampoline.  Also, I really don’t like mixed hex and octal in an address map – it’s a recipe for confusion.

Reply:
haven't tried views yet, will leave as is for now.   map will use octal only, along with other parts of the code.

Reply:
The map for your bank device is a simple case that would be simpler to read with a view in the memory map.  It’s not doing anything that’s difficult to do with views (e.g. using mirrors to install something across multiple banks, or installing the bank device in multiple locations).  It’s really better not to add legacy code.